README.md
# UK Car Tax calculator

A node module to calculate vehicle tax rates for cars and vans in the UK based on the fuel type, engine size and when the vehicle was registered.

[![Build Status](https://dev.azure.com/supermarioryan/UK%20Car%20Tax%20Calculator/_apis/build/status/rymate1234.uk-car-tax-calculator?branchName=master)](https://dev.azure.com/supermarioryan/UK%20Car%20Tax%20Calculator/_build/latest?definitionId=3&branchName=master)

## Installation

```
npm install uk-car-tax-calculator
```

## Usage

```js
const calculateTax = require('uk-car-tax-calculator') // or es modules

const tax = calculateTax({
  registrationDate: new Date(2001, 3, 1),
  co2: 201,
  fuel: 'Hybrid'
})

console.log(tax.price) // prints 315
```

For a more comprehensive guide to usage check `test/index.test.js`

## Contributing

PRs accepted, especially those concerning updates to UK law.

This project is linted with `eslint` using prettier, and runs tests with `jest` - these will need to pass before any PR is accepted.


CHANGELOG.md
# 2.0.0

This major release adds support for returning the tax category and tax band (if applicable)

This is a BREAKING CHANGE! The returned object is now in the following format:

```ts
interface VehicleTax {
    price: number;
    band?: string;
    category?: string;
}
```

# 1.1.0

This version adds support for calculating the tax rate of vans / light goods vehicles

# 1.0.1

More sensible build defaults for use in environments where transpiling takes place:

- For ES Modules the src/index.js is directly referenced rather than a minified version
- Microbundle is configured to concatenate all files into one for the cjs builds, and minify a unpkg mjs file and umd js file for browser / non node usage

# 1.0.0

Initial Version


test/index.test.js
const calc = require('../dist')

describe('Historic Vehicles', () => {
  const registrationDate = new Date(1978, 1, 1)

  test('should return 0 when calculating a car before 1979-01-01', () => {
    expect(calc({ registrationDate }).price).toBe(0)
  })
})

describe('Electric Vehicles', () => {
  const registrationDate = new Date(2018, 3, 1)

  test('should return 0 when an electric car', () => {
    expect(calc({ registrationDate, co2: 123, fuel: 'Electric' }).price).toBe(0)
  })

  test('should return 320 when an electric car with a value more than £40000', () => {
    expect(
      calc({ registrationDate, co2: 123, fuel: 'Electric', value: 45000 }).price
    ).toBe(320)
  })
})

describe('Vans', () => {
  test('should return 260 when registered on or after 1 March 2001', () => {
    const registrationDate = new Date(2018, 3, 1)
    const taxInfo = calc({ registrationDate, co2: 123, type: 'Van' })
    expect(taxInfo.price).toBe(260)
  })

  test('should return 260 when registered before 1 March 2001', () => {
    const registrationDate = new Date(1979, 1, 1)
    expect(
      calc({ registrationDate, co2: 123, type: 'Van', engineSize: 1549 }).price
    ).toBe(155)
  })

  test('should return 140 for a euro 4 compliant van', () => {
    const registrationDate = new Date(2004, 1, 1)
    expect(
      calc({ registrationDate, co2: 123, type: 'Van', euroStandard: 4 }).price
    ).toBe(140)
  })

  test('should return 140 for a euro 5 compliant van', () => {
    const registrationDate = new Date(2009, 1, 1)
    expect(
      calc({ registrationDate, co2: 123, type: 'Van', euroStandard: 5 }).price
    ).toBe(140)
  })

  test('should return 0 for an electric van', () => {
    const registrationDate = new Date(2018, 3, 1)
    expect(
      calc({ registrationDate, co2: 123, type: 'Van', fuel: 'Electric' }).price
    ).toBe(0)
  })
})

describe('Registered before 1 March 2001', () => {
  const registrationDate = new Date(1979, 1, 1)

  test('should return 155 with an engine size of 1549 or less', () => {
    expect(calc({ registrationDate, engineSize: 1549 }).price).toBe(155)
    expect(calc({ registrationDate, engineSize: 1349 }).price).toBe(155)
  })

  test('should return 255 with an engine size of greater than 1549', () => {
    expect(calc({ registrationDate, engineSize: 1599 }).price).toBe(255)
  })

  test('should return the correct tax category', () => {
    expect(calc({ registrationDate, engineSize: 1599 }).category).toBe('TC11')
  })
})

describe('Registered between 1 March 2001 and 31 March 2017', () => {
  const registrationDate = new Date(2001, 3, 1)
  const date2006 = new Date(2006, 3, 1)

  test('should return 0 with CO2 emissions of 100 or less', () => {
    expect(calc({ registrationDate, co2: 0 }).price).toBe(0)
    expect(calc({ registrationDate, co2: 100 }).price).toBe(0)
  })

  test('should return 20 with CO2 emissions between 100 and 110 or less', () => {
    expect(calc({ registrationDate, co2: 101, fuel: 'Petrol' }).price).toBe(20)
    expect(calc({ registrationDate, co2: 101, fuel: 'Diesel' }).price).toBe(20)
  })

  test('should return 325 with CO2 emissions 201 or above before 23rd march 2006', () => {
    expect(calc({ registrationDate, co2: 201, fuel: 'Petrol' }).price).toBe(325)
    expect(calc({ registrationDate, co2: 255, fuel: 'Diesel' }).price).toBe(325)
    expect(calc({ registrationDate, co2: 256, fuel: 'Diesel' }).price).toBe(325)
  })

  test('should return 325 with CO2 emissions 201-225 or above after 23rd march 2006', () => {
    expect(
      calc({ registrationDate: date2006, co2: 201, fuel: 'Petrol' }).price
    ).toBe(325)
    expect(
      calc({ registrationDate: date2006, co2: 255, fuel: 'Diesel' }).price
    ).toBe(555)
    expect(
      calc({ registrationDate: date2006, co2: 256, fuel: 'Diesel' }).price
    ).toBe(570)
  })

  test('should return 10 less than petrol / diesel with alternative fuels', () => {
    expect(calc({ registrationDate, co2: 101, fuel: 'Hybrid' }).price).toBe(10)
    expect(calc({ registrationDate, co2: 201, fuel: 'Hybrid' }).price).toBe(315)
    expect(calc({ registrationDate, co2: 255, fuel: 'Hybrid' }).price).toBe(315)
    expect(calc({ registrationDate, co2: 256, fuel: 'Hybrid' }).price).toBe(315)

    expect(
      calc({ registrationDate: date2006, co2: 201, fuel: 'Hybrid' }).price
    ).toBe(315)
    expect(
      calc({ registrationDate: date2006, co2: 255, fuel: 'Hybrid' }).price
    ).toBe(545)
    expect(
      calc({ registrationDate: date2006, co2: 256, fuel: 'Hybrid' }).price
    ).toBe(560)
  })

  test('should return the correct tax band', () => {
    expect(calc({ registrationDate, co2: 256, fuel: 'Petrol' }).band).toBe('K')
    expect(calc({ registrationDate, co2: 256, fuel: 'Diesel' }).band).toBe('K')
    expect(calc({ registrationDate, co2: 256, fuel: 'Hybrid' }).band).toBe('K')

    expect(
      calc({ registrationDate: date2006, co2: 256, fuel: 'Petrol' }).band
    ).toBe('M')
    expect(
      calc({ registrationDate: date2006, co2: 256, fuel: 'Diesel' }).band
    ).toBe('M')
    expect(
      calc({ registrationDate: date2006, co2: 256, fuel: 'Hybrid' }).band
    ).toBe('M')
  })
})

describe('Registered on or after 1 April 2017', () => {
  const registrationDate = new Date(2018, 3, 1)

  test('should return 0 with CO2 emissions of 0', () => {
    expect(calc({ registrationDate, co2: 0, fuel: 'Petrol' }).price).toBe(0)
    expect(calc({ registrationDate, co2: 1, fuel: 'Petrol' }).price).not.toBe(0)
  })

  test('should return 0 with CO2 emissions of 50 or less for hybrids', () => {
    expect(calc({ registrationDate, co2: 0, fuel: 'Hybrid' }).price).toBe(0)
    expect(calc({ registrationDate, co2: 49, fuel: 'Hybrid' }).price).toBe(0)
    expect(calc({ registrationDate, co2: 51, fuel: 'Hybrid' }).price).not.toBe(
      0
    )
  })

  test('should return the right value for CO2 emissions of 1 - 50', () => {
    expect(calc({ registrationDate, co2: 25, fuel: 'Petrol' }).price).toBe(10)
    expect(calc({ registrationDate, co2: 25, fuel: 'Diesel' }).price).toBe(25)

    expect(
      calc({ registrationDate, co2: 25, fuel: 'Diesel', meetsRDE2: true }).price
    ).toBe(10)
  })

  test('should return the right value for CO2 emissions of 226 - 255', () => {
    expect(calc({ registrationDate, co2: 230, fuel: 'Petrol' }).price).toBe(
      1815
    )
    expect(calc({ registrationDate, co2: 230, fuel: 'Diesel' }).price).toBe(
      2135
    )
    expect(calc({ registrationDate, co2: 230, fuel: 'Hybrid' }).price).toBe(
      1805
    )

    expect(
      calc({ registrationDate, co2: 230, fuel: 'Diesel', meetsRDE2: true })
        .price
    ).toBe(1815)
  })

  test('should return the right value for max CO2 emissions (>255)', () => {
    expect(calc({ registrationDate, co2: 256, fuel: 'Petrol' }).price).toBe(
      2135
    )
    expect(calc({ registrationDate, co2: 256, fuel: 'Diesel' }).price).toBe(
      2135
    )
    expect(calc({ registrationDate, co2: 256, fuel: 'Hybrid' }).price).toBe(
      2125
    )

    expect(
      calc({ registrationDate, co2: 256, fuel: 'Diesel', meetsRDE2: true })
        .price
    ).toBe(2135)
  })
})

describe('Tax Categories', () => {
  const registrationDate = new Date(2018, 3, 1)

  test('should return TC48 for petrol', () => {
    expect(calc({ registrationDate, fuel: 'Petrol' }).category).toBe('TC48')
  })

  test('should return TC49 for diesel', () => {
    expect(calc({ registrationDate, fuel: 'Diesel' }).category).toBe('TC49')
  })

  test('should return TC59 for alternative fuels', () => {
    expect(calc({ registrationDate, fuel: 'Hybrid' }).category).toBe('TC59')
  })
})


src/index.js
import { getPre2017Mappings, getMappings } from './mappings'

const HistoricVehicle = new Date(1979, 0, 1)
const PreCO2 = new Date(2001, 2, 1)
const April2017 = new Date(2017, 3, 1)

const Euro4Start = new Date(2003, 2, 1)
const Euro4End = new Date(2006, 11, 31)

const Euro5Start = new Date(2009, 0, 1)
const Euro5End = new Date(2010, 11, 31)

/**
 * Calculate tax based on engine size
 *
 * @param {number} engineSize Size of the engine
 * @returns {number} The tax amount
 */
function calculateEngineSizeCarTax(engineSize) {
  if (engineSize <= 1549) {
    return 155
  } else {
    return 255
  }
}

/**
 * Reduces tax by £10 if it's not a petrol / diesel car
 *
 * @param {number} tax Current tax amount
 * @param {string} fuel The fuel used
 * @returns {number} The final tax amount
 */
function reduceIfAlternative(tax, fuel) {
  if (fuel !== 'petrol' && fuel !== 'diesel') {
    return tax > 0 ? tax - 10 : 0
  }

  return tax
}

/**
 * Derives the tax category from the fuel of the vehicle
 *
 * @param {string} fuel The fuel used
 * @returns {string} The tax category
 */
function fuelToTaxCategory(fuel) {
  if (fuel === 'petrol') {
    return 'TC48'
  }

  if (fuel === 'diesel') {
    return 'TC49'
  }

  if (fuel !== 'petrol' && fuel !== 'diesel') {
    return 'TC59'
  }
}

/**
 * Calculates car tax for vehicles registered between 1 March 2001 and 31 March 2017
 *
 * @param {Date} registrationDate When the vehicle was registered
 * @param {number} co2 Amount of CO2 produced
 * @param {string} fuel The fuel used
 *
 * @returns {number} The final tax amount
 */
function calculatePre2017CarTax(registrationDate, co2, fuel) {
  const tax = getPre2017Mappings(registrationDate, co2)
  tax.price = reduceIfAlternative(tax.price, fuel)
  return tax
}

/**
 * Calculates car tax for vehicles registered on or after 1 April 2017
 *
 * @param {number} co2 Amount of CO2 produced
 * @param {string} fuel The fuel used
 * @param {boolean} meetsRDE2 Whether the vehicle meets RDE2
 * @returns {number} The final tax amount
 */
function calculateCurrentCarTax(co2, fuel, meetsRDE2) {
  if (fuel != 'diesel') {
    meetsRDE2 = true
  }
  const tax = getMappings(co2, meetsRDE2)
  tax.price = reduceIfAlternative(tax.price, fuel)
  return tax
}

/**
 * Calculate the current car tax of a vehicle
 * @param {VehicleInfo} options The list of vehicle options
 * @returns {number} The final tax amount
 */
function calculateTax(options) {
  let {
    registrationDate,
    engineSize,
    co2,
    fuel = '',
    meetsRDE2,
    value,
    type = '',
    euroStandard
  } = options

  const taxResult = {
    price: 0,
    category: '',
    band: ''
  }

  if (registrationDate < HistoricVehicle) {
    return taxResult
  }

  fuel = fuel.toLowerCase()
  type = type.toLowerCase()

  taxResult.category = fuelToTaxCategory(fuel)

  const isCurrent = registrationDate >= April2017

  if (fuel === 'electric') {
    taxResult.price = value > 40000 && isCurrent ? 320 : 0
    return taxResult
  }

  if (registrationDate < PreCO2) {
    taxResult.price = calculateEngineSizeCarTax(engineSize)
    taxResult.category = 'TC11'
    return taxResult
  }

  if (registrationDate >= PreCO2 && type === 'van') {
    const euro4Van =
      registrationDate >= Euro4Start &&
      registrationDate < Euro4End &&
      euroStandard === 4

    const euro5Van =
      registrationDate >= Euro5Start &&
      registrationDate < Euro5End &&
      euroStandard === 5

    if (euro4Van || euro5Van) {
      taxResult.price = 140
      taxResult.category = 'TC36'
    } else {
      taxResult.price = 260
      taxResult.category = 'TC39'
    }

    return taxResult
  }

  if (registrationDate >= PreCO2 && registrationDate < April2017) {
    return Object.assign(
      {},
      taxResult,
      calculatePre2017CarTax(registrationDate, co2, fuel)
    )
  }

  if (isCurrent) {
    return Object.assign(
      {},
      taxResult,
      calculateCurrentCarTax(co2, fuel, meetsRDE2)
    )
  }
}

export default calculateTax


src/mappings.js
/**
 * Calculates car tax for vehicles registered between 1 March 2001 and 31 March 2017
 * @param {Date} registrationDate When the vehicle was registered
 * @param {number} co2 Amount of CO2 produced
 */
export function getPre2017Mappings(registrationDate, co2) {
  const pre2006 = new Date(2006, 2, 23)
  const values = [
    { co2: 100, band: 'A', price: 0 },
    { co2: 110, band: 'B', price: 20 },
    { co2: 120, band: 'C', price: 30 },
    { co2: 130, band: 'D', price: 125 },
    { co2: 140, band: 'E', price: 145 },
    { co2: 150, band: 'F', price: 160 },
    { co2: 165, band: 'G', price: 200 },
    { co2: 175, band: 'H', price: 235 },
    { co2: 185, band: 'I', price: 260 },
    { co2: 200, band: 'J', price: 300 }
  ]

  for (const value of values) {
    const { price, band } = value
    if (co2 <= value.co2) return { price, band }
  }

  if (co2 <= 225 || (co2 > 226 && pre2006 > registrationDate))
    return { band: 'K', price: 325 }

  if (co2 <= 255) return { band: 'L', price: 555 }
  if (co2 > 255) return { band: 'M', price: 570 }
}

/**
 * Calculates car tax for vehicles registered after April 2017
 * @param {number} co2 Amount of CO2 produced
 * @param {boolean} meetsRDE2 Whether the vehicle meets RDE2
 */
export function getMappings(co2, meetsRDE2) {
  const taxes = [0, 10, 25, 110, 130, 150, 170, 210, 530, 855, 1280, 1815, 2135]
  const thresholds = [0, 50, 75, 90, 100, 110, 130, 150, 170, 190, 225, 255]
  let price = 0
  if (!meetsRDE2) {
    taxes.splice(1, 1)
    taxes.push(taxes[taxes.length - 1])
  }

  for (const threshold of thresholds) {
    price = taxes.shift()
    if (co2 <= threshold) break
  }

  if (co2 > 255) price = taxes.shift()

  return { price }
}
